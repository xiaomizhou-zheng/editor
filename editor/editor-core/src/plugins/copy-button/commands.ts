import { Command, CommandDispatch } from '../../types';
import { copyButtonPluginKey } from './pm-plugins/plugin-key';
import { MarkType, NodeType } from 'prosemirror-model';
import { hoverDecoration } from '../base/pm-plugins/decoration';
import { EditorState, Transaction } from 'prosemirror-state';
import { copyHTMLToClipboard } from '../../utils/clipboard';
import { getSelectedNodeOrNodeParentByNodeType, toDOM } from './utils';
import { addAnalytics, ACTION, INPUT_METHOD } from '../analytics';
import { getAnalyticsPayload } from '../clipboard/pm-plugins/main';

export function createToolbarCopyCommandForMark(markType: MarkType): Command {
  function command(state: EditorState, dispatch: CommandDispatch | undefined) {
    const textNode = state.tr.selection.$head.parent.maybeChild(
      state.tr.selection.$head.index(),
    );

    if (!textNode) {
      return false;
    }

    if (dispatch) {
      // As calling copyHTMLToClipboard causes side effects -- we only run this when
      // dispatch is provided -- as otherwise the consumer is only testing to see if
      // the action is availble.
      const domNode = toDOM(textNode, state.schema);
      if (domNode) {
        const div = document.createElement('div');
        const p = document.createElement('p');
        div.appendChild(p);
        p.appendChild(domNode);
        // The "1 1" refers to the start and end depth of the slice
        // since we're copying the text inside a paragraph, it will always be 1 1
        // https://github.com/ProseMirror/prosemirror-view/blob/master/src/clipboard.ts#L32
        (div.firstChild as HTMLElement).setAttribute('data-pm-slice', '1 1 []');

        copyHTMLToClipboard(div.innerHTML);
      }

      const copyToClipboardTr = state.tr;
      copyToClipboardTr.setMeta(copyButtonPluginKey, { copied: true });

      const analyticsPayload = getAnalyticsPayload(state, ACTION.COPIED);
      if (analyticsPayload) {
        analyticsPayload.attributes.inputMethod = INPUT_METHOD.FLOATING_TB;
        analyticsPayload.attributes.markType = markType.name;
        addAnalytics(state, copyToClipboardTr, analyticsPayload);
      }

      dispatch(copyToClipboardTr);
    }

    return true;
  }

  return command;
}

export function getProvideMarkVisualFeedbackForCopyButtonCommand(
  markType: MarkType,
) {
  function provideMarkVisualFeedbackForCopyButtonCommand(
    state: EditorState,
    dispatch: CommandDispatch | undefined,
  ) {
    const tr = state.tr;
    tr.setMeta(copyButtonPluginKey, { showSelection: true, markType });

    if (dispatch) {
      dispatch(tr);
    }

    return true;
  }
  return provideMarkVisualFeedbackForCopyButtonCommand;
}

export function removeMarkVisualFeedbackForCopyButtonCommand(
  state: EditorState,
  dispatch: CommandDispatch | undefined,
) {
  const tr = state.tr;
  tr.setMeta(copyButtonPluginKey, { removeSelection: true });

  const copyButtonState = copyButtonPluginKey.getState(state);
  if (copyButtonState?.copied) {
    tr.setMeta(copyButtonPluginKey, { copied: false });
  }

  if (dispatch) {
    dispatch(tr);
  }

  return true;
}

export const createToolbarCopyCommandForNode = (
  nodeType: NodeType | Array<NodeType>,
): Command => (state, dispatch) => {
  const { tr, schema } = state;

  // This command should only be triggered by the Copy button in the floating toolbar
  // which is only visible when selection is inside the target node
  let contentNodeWithPos = getSelectedNodeOrNodeParentByNodeType({
    nodeType,
    selection: tr.selection,
  });
  if (!contentNodeWithPos) {
    return false;
  }

  const copyToClipboardTr = tr;
  copyToClipboardTr.setMeta(copyButtonPluginKey, { copied: true });

  const analyticsPayload = getAnalyticsPayload(state, ACTION.COPIED);
  if (analyticsPayload) {
    analyticsPayload.attributes.inputMethod = INPUT_METHOD.FLOATING_TB;
    analyticsPayload.attributes.nodeType = contentNodeWithPos.node.type.name;
    addAnalytics(state, copyToClipboardTr, analyticsPayload);
  }
  if (dispatch) {
    // As calling copyHTMLToClipboard causes side effects -- we only run this when
    // dispatch is provided -- as otherwise the consumer is only testing to see if
    // the action is availble.
    const domNode = toDOM(contentNodeWithPos.node, schema);
    if (domNode) {
      const div = document.createElement('div');
      div.appendChild(domNode);

      // if copying inline content
      if (contentNodeWithPos.node.type.inlineContent) {
        // The "1 1" refers to the start and end depth of the slice
        // since we're copying the text inside a paragraph, it will always be 1 1
        // https://github.com/ProseMirror/prosemirror-view/blob/master/src/clipboard.ts#L32
        (div.firstChild as HTMLElement).setAttribute('data-pm-slice', '1 1 []');
      } else {
        // The "0 0" refers to the start and end depth of the slice
        // since we're copying the block node only, it will always be 0 0
        // https://github.com/ProseMirror/prosemirror-view/blob/master/src/clipboard.ts#L32
        (div.firstChild as HTMLElement).setAttribute('data-pm-slice', '0 0 []');
      }
      copyHTMLToClipboard(div.innerHTML);
    }
    copyToClipboardTr.setMeta('scrollIntoView', false);
    dispatch(copyToClipboardTr);
  }

  return true;
};

export const resetCopiedState = (
  nodeType: NodeType | Array<NodeType>,
  onMouseLeave?: Command,
): Command => (state, dispatch) => {
  let customTr = state.tr;

  // Avoid multipe dispatch
  // https://product-fabric.atlassian.net/wiki/spaces/E/pages/2241659456/All+about+dispatch+and+why+there+shouldn+t+be+multiple#How-do-I-avoid-them%3F
  const customDispatch = (tr: Transaction) => {
    customTr = tr;
  };

  onMouseLeave
    ? onMouseLeave(state, customDispatch)
    : hoverDecoration(nodeType, false)(state, customDispatch);

  const copyButtonState = copyButtonPluginKey.getState(state);
  if (copyButtonState?.copied) {
    customTr.setMeta(copyButtonPluginKey, { copied: false });
  }

  if (dispatch) {
    dispatch(customTr);
  }

  return true;
};
