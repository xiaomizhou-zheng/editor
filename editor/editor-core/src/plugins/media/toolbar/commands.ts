import { Fragment } from 'prosemirror-model';
import { NodeSelection } from 'prosemirror-state';
import {
  isNodeSelection,
  removeSelectedNode,
  safeInsert,
} from 'prosemirror-utils';
import { Command } from '../../../types';
import {
  ACTION,
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
  addAnalytics,
  EVENT_TYPE,
} from '../../analytics';
import { removeMediaGroupNode } from './utils';

export const changeInlineToMediaCard: Command = (state, dispatch) => {
  const { media, mediaInline, mediaGroup } = state.schema.nodes;
  const selectedNode =
    state.selection instanceof NodeSelection &&
    state.selection.node.type === mediaInline &&
    state.selection.node;

  if (!selectedNode) {
    return false;
  }

  const { id, type, collection } = selectedNode.attrs;
  const mediaNode = media.createChecked({
    id,
    type,
    collection,
  });
  const group = mediaGroup.createChecked({}, mediaNode);

  const nodePos = state.tr.doc.resolve(state.selection.from).end();

  let tr = state.tr;
  tr = removeSelectedNode(tr);
  tr = safeInsert(group, nodePos, true)(tr);

  if (dispatch) {
    addAnalytics(state, tr, {
      action: ACTION.CHANGED_TYPE,
      actionSubject: ACTION_SUBJECT.MEDIA,
      eventType: EVENT_TYPE.TRACK,
      attributes: {
        newType: ACTION_SUBJECT_ID.MEDIA_GROUP,
        previousType: ACTION_SUBJECT_ID.MEDIA_INLINE,
      },
    });

    dispatch(tr);
  }

  return true;
};

export const changeMediaCardToInline: Command = (state, dispatch) => {
  const { media, mediaInline, paragraph } = state.schema.nodes;
  const selectedNode =
    state.selection instanceof NodeSelection && state.selection.node;

  if (!selectedNode || !selectedNode.type === media) {
    return false;
  }

  const mediaInlineNode = mediaInline.create({
    id: selectedNode.attrs.id,
    collection: selectedNode.attrs.collection,
  });
  const space = state.schema.text(' ');
  let content = Fragment.from([mediaInlineNode, space]);
  const node = paragraph.createChecked({}, content);

  const nodePos = state.tr.doc.resolve(state.selection.from).start() - 1;

  let tr = removeMediaGroupNode(state);
  tr = safeInsert(node, nodePos, true)(tr);

  if (dispatch) {
    addAnalytics(state, tr, {
      action: ACTION.CHANGED_TYPE,
      actionSubject: ACTION_SUBJECT.MEDIA,
      eventType: EVENT_TYPE.TRACK,
      attributes: {
        newType: ACTION_SUBJECT_ID.MEDIA_INLINE,
        previousType: ACTION_SUBJECT_ID.MEDIA_GROUP,
      },
    });
    dispatch(tr);
  }

  return true;
};

export const removeInlineCard: Command = (state, dispatch) => {
  if (isNodeSelection(state.selection)) {
    if (dispatch) {
      dispatch(removeSelectedNode(state.tr));
    }
    return true;
  }
  return false;
};
