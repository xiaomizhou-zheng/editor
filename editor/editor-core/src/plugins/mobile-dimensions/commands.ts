import { MobileDimensionsActionTypes } from './actions';
import { createCommand } from './plugin-factory';

export const setKeyboardHeight = (keyboardHeight: number) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_KEYBOARD_HEIGHT,
    keyboardHeight,
  });

export const setWindowHeight = (windowHeight: number) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_WINDOW_HEIGHT,
    windowHeight,
  });

export const setMobilePaddingTop = (paddingTop: number) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_MOBILE_PADDING_TOP,
    paddingTop,
  });

export const setIsExpanded = (isExpanded: boolean) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_IS_EXPANDED,
    isExpanded,
  });
