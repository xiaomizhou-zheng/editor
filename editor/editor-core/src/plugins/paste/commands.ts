import { createCommand } from './pm-plugins/plugin-factory';
import { PastePluginActionTypes as ActionTypes } from './actions';

/**
 * Use this to register macro link positions during a paste operation, that you
 * want to track in a document over time, through any document changes.
 *
 * @param positions a map of string keys (custom position references) and position values e.g. { ['my-key-1']: 11 }
 *
 * **Context**: This is neccessary if there is an async process or an unknown period of time
 * between obtaining an original position, and wanting to know about what its final eventual
 * value. In that scenario, positions will need to be actively tracked and mapped in plugin
 * state so that they can be mapped through any other independent document change transactions being
 * dispatched to the editor that could affect their value.
 */
export const startTrackingPastedMacroPositions = (pastedMacroPositions: {
  [key: string]: number;
}) =>
  createCommand(() => {
    return {
      type: ActionTypes.START_TRACKING_PASTED_MACRO_POSITIONS,
      pastedMacroPositions,
    };
  });

export const stopTrackingPastedMacroPositions = (
  pastedMacroPositionKeys: string[],
) =>
  createCommand(() => {
    return {
      type: ActionTypes.STOP_TRACKING_PASTED_MACRO_POSITIONS,
      pastedMacroPositionKeys,
    };
  });
