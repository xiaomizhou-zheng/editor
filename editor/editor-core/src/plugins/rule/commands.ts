import { Command } from '../../types';
import { INPUT_METHOD } from '../analytics';
import { createHorizontalRule } from './pm-plugins/input-rule';

export const insertHorizontalRule = (
  inputMethod:
    | INPUT_METHOD.QUICK_INSERT
    | INPUT_METHOD.TOOLBAR
    | INPUT_METHOD.INSERT_MENU
    | INPUT_METHOD.FORMATTING
    | INPUT_METHOD.SHORTCUT,
): Command => (state, dispatch) => {
  const tr = createHorizontalRule(
    state,
    state.selection.from,
    state.selection.to,
    inputMethod,
  );
  if (tr) {
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  }

  return false;
};
