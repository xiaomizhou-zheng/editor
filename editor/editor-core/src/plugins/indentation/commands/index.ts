import { Node as PmNode, Schema } from 'prosemirror-model';
import { IndentationMarkAttributes } from '@atlaskit/adf-schema';
import { toggleBlockMark } from '../../../commands';
import { Command } from '../../../types/command';
import { createAnalyticsDispatch, IndentationInputMethod } from './utils';
import { INDENT_DIRECTION, INPUT_METHOD } from '../../analytics';
import getAttrsWithChangesRecorder from '../../../utils/getAttrsWithChangesRecorder';

export const MAX_INDENTATION_LEVEL = 6;

export const isIndentationAllowed = (schema: Schema, node: PmNode) => {
  const {
    nodes: { paragraph, heading },
    marks: { alignment },
  } = schema;

  if ([paragraph, heading].indexOf(node.type) > -1) {
    if (alignment) {
      const hasAlignment = node.marks.filter(
        (mark) => mark.type === alignment,
      )[0];
      return !hasAlignment;
    }
    return true;
  }
  return false;
};

/**
 * Create new indentation command (Either indent or outdent depend of getArgsFn)
 * @param getNewIndentationAttrs Function to handle new indentation level
 */
function createIndentationCommand(
  getNewIndentationAttrs: (
    prevArgs?: IndentationMarkAttributes,
    node?: PmNode,
  ) => IndentationMarkAttributes | undefined | false,
): Command {
  return (state, dispatch) => {
    const { indentation } = state.schema.marks;

    return toggleBlockMark<IndentationMarkAttributes>(
      indentation,
      getNewIndentationAttrs,
      isIndentationAllowed,
    )(state, dispatch);
  };
}

function createIndentationCommandWithAnalytics({
  getNewIndentationAttrs,
  direction,
  inputMethod,
}: {
  getNewIndentationAttrs: (
    prevAttrs?: IndentationMarkAttributes,
    node?: PmNode,
  ) => IndentationMarkAttributes | undefined | false;
  direction: INDENT_DIRECTION;
  inputMethod: IndentationInputMethod;
}): Command {
  // Create a new getAttrs function to record the changes
  const {
    getAttrs,
    getAndResetAttrsChanges,
  } = getAttrsWithChangesRecorder(getNewIndentationAttrs, { direction });

  // Use new getAttrs wrapper
  const indentationCommand = createIndentationCommand(getAttrs);

  // Return a new command where we change dispatch for our analytics dispatch
  return (state, dispatch) => {
    return indentationCommand(
      state,
      createAnalyticsDispatch({
        getAttrsChanges: getAndResetAttrsChanges,
        inputMethod,
        state,
        dispatch,
      }),
    );
  };
}

/**
 * Get new level for outdent
 * @param oldAttr Old attributes for the mark, undefined if the mark doesn't exit
 * @returns  - undefined; No change required
 *           - false; Remove the mark
 *           - object; Update attributes
 */
const getIndentAttrs = (
  oldAttr?: IndentationMarkAttributes,
): IndentationMarkAttributes | undefined => {
  if (!oldAttr) {
    return { level: 1 }; // No mark exist, create a new one with level 1
  }

  const { level } = oldAttr;
  if (level >= MAX_INDENTATION_LEVEL) {
    return undefined; // Max indentation level reached, do nothing.
  }
  return { level: level + 1 }; // Otherwise, increase the level by one
};

export const getIndentCommand = (
  inputMethod: IndentationInputMethod = INPUT_METHOD.KEYBOARD,
): Command =>
  createIndentationCommandWithAnalytics({
    getNewIndentationAttrs: getIndentAttrs,
    direction: INDENT_DIRECTION.INDENT,
    inputMethod,
  });

/**
 * Get new level for outdent
 * @param oldAttr Old attributes for the mark, undefined if the mark doesn't exit
 * @returns  - undefined; No change required
 *           - false; Remove the mark
 *           - object; Update attributes
 */
const getOutdentAttrs = (
  oldAttr?: IndentationMarkAttributes,
): IndentationMarkAttributes | undefined | false => {
  if (!oldAttr) {
    return undefined; // Do nothing;
  }

  const { level } = oldAttr;
  if (level <= 1) {
    return false; // Remove the mark
  }

  return { level: level - 1 }; // Decrease the level on other cases
};

export const getOutdentCommand = (
  inputMethod: IndentationInputMethod = INPUT_METHOD.KEYBOARD,
): Command =>
  createIndentationCommandWithAnalytics({
    getNewIndentationAttrs: getOutdentAttrs,
    direction: INDENT_DIRECTION.OUTDENT,
    inputMethod,
  });

export const removeIndentation: Command = (state, dispatch) =>
  toggleBlockMark(state.schema.marks.indentation, () => false)(state, dispatch);
