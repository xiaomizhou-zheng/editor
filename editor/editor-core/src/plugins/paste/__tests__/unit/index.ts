import {
  Preset,
  LightEditorPlugin,
  createProsemirrorEditorFactory,
} from '@atlaskit/editor-test-helpers/create-prosemirror-editor';
import dispatchPasteEvent from '@atlaskit/editor-test-helpers/dispatch-paste-event';
import { EditorView } from 'prosemirror-view';
import { toggleStrong } from '../../../text-formatting/commands/text-formatting';
import pastePlugin from '../../index';
import blockTypePlugin from '../../../block-type';
import { textFormattingPlugin } from '../../../index';
import hyperlinkPlugin from '../../../hyperlink/index';
import { createEditorFactory } from '@atlaskit/editor-test-helpers/create-editor';
import {
  doc,
  p,
  panel,
  taskList,
  taskItem,
  decisionList,
  decisionItem,
  DocBuilder,
} from '@atlaskit/editor-test-helpers/doc-builder';

describe('Paste plugin', () => {
  const createEditor = createProsemirrorEditorFactory();
  let editorView: EditorView;

  const editor = (doc: DocBuilder, plainTextPasteLinkification: boolean) =>
    createEditor({
      doc,
      preset: new Preset<LightEditorPlugin>()
        .add([pastePlugin, { plainTextPasteLinkification }])
        .add(blockTypePlugin)
        .add(hyperlinkPlugin)
        .add(textFormattingPlugin),
    });

  describe('With plainTextPasteLinkification FF enabled', () => {
    beforeEach(() => {
      ({ editorView } = editor(doc(p('{<>}')), true));
    });

    describe('cmd+shift+v', () => {
      describe('content without any links', () => {
        const paste = () => {
          dispatchPasteEvent(
            editorView,
            {
              html: `<meta charset='utf-8'><div style="color: #333333;background-color: #f5f5f5;font-family: Menlo, Monaco, 'Courier New', monospace;font-weight: normal;font-size: 12px;line-height: 18px;white-space: pre;"><div><span style="color: #4b69c6;">export</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;">interface</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;font-weight: bold;">PasteContent</span></div></div>`,
              plain: 'export interface PasteContent',
            },
            { shift: true },
          );
        };

        it('pastes plain text (ie. removes formatting from pasted content)', () => {
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });

        it('preserves current formatting when pasting (ie. removes formatting, applies active formatting)', () => {
          toggleStrong()(editorView.state, editorView.dispatch);
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });
      });

      describe('content with link', () => {
        const paste = () => {
          dispatchPasteEvent(
            editorView,
            {
              html: `<meta charset='utf-8'><div style="color: #333333;background-color: #f5f5f5;font-family: Menlo, Monaco, 'Courier New', monospace;font-weight: normal;font-size: 12px;line-height: 18px;white-space: pre;"><div><span style="color: #4b69c6;">export</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;">interface</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;font-weight: bold;">PasteContent</span> <a href="https://www.google.com">https://www.google.com</a> <span>some more content</span></div></div>`,
              plain:
                'export interface PasteContent https://www.google.com some more content',
            },
            { shift: true },
          );
        };
        it('pastes plain text but creates hyperlink (ie. removes formatting from pasted content)', () => {
          // This does not test that it doesn't create a *smart* link.
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });

        it('preserves current formatting when pasting, creates hyperlink (ie. removes formatting, applies active formatting)', () => {
          // This does not test that it doesn't create a *smart* link.
          toggleStrong()(editorView.state, editorView.dispatch);
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });
      });
    });
  });

  describe('With plainTextPasteLinkification FF disabled', () => {
    beforeEach(() => {
      ({ editorView } = editor(doc(p('{<>}')), false));
    });
    describe('cmd+shift+v', () => {
      describe('content without any links', () => {
        const paste = () => {
          dispatchPasteEvent(
            editorView,
            {
              html: `<meta charset='utf-8'><div style="color: #333333;background-color: #f5f5f5;font-family: Menlo, Monaco, 'Courier New', monospace;font-weight: normal;font-size: 12px;line-height: 18px;white-space: pre;"><div><span style="color: #4b69c6;">export</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;">interface</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;font-weight: bold;">PasteContent</span></div></div>`,
              plain: 'export interface PasteContent',
            },
            { shift: true },
          );
        };

        it('pastes plain text (ie. removes formatting from pasted content)', () => {
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });

        it('preserves current formatting when pasting (ie. removes formatting, applies active formatting)', () => {
          toggleStrong()(editorView.state, editorView.dispatch);
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });
      });

      describe('content with link', () => {
        const paste = () => {
          dispatchPasteEvent(
            editorView,
            {
              html: `<meta charset='utf-8'><div style="color: #333333;background-color: #f5f5f5;font-family: Menlo, Monaco, 'Courier New', monospace;font-weight: normal;font-size: 12px;line-height: 18px;white-space: pre;"><div><span style="color: #4b69c6;">export</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;">interface</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;font-weight: bold;">PasteContent</span> <a href="https://www.google.com">https://www.google.com</a> <span>some more content</span></div></div>`,
              plain:
                'export interface PasteContent https://www.google.com some more content',
            },
            { shift: true },
          );
        };
        it("pastes plain text and doesn't create hyperlink", () => {
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });

        it(`preserves current formatting when pasting, doesn't create hyperlink (ie. removes formatting, applies active formatting)`, () => {
          // This does not test that it doesn't create a *smart* link.
          toggleStrong()(editorView.state, editorView.dispatch);
          paste();
          expect(editorView.state.doc).toMatchSnapshot();
        });
      });
    });
  });

  describe('when pasting content copied from the top level of the document (openStart is 0)', () => {
    describe('into a panel', () => {
      beforeEach(() => {
        const createEditor = createEditorFactory();
        const editor = (doc: DocBuilder) =>
          createEditor({
            doc,
            editorProps: {
              allowPanel: {
                allowCustomPanel: true,
                allowCustomPanelEdit: true,
              },
              allowTasksAndDecisions: true,
            },
          });

        ({ editorView } = editor(doc(panel()(p('{<>}')))));
      });

      it('the panel should not disappear when pasting one paragraph', () => {
        dispatchPasteEvent(editorView, {
          html: `<meta charset='utf-8'><p data-pm-slice="0 0 []">abc</p>`,
          plain: 'export interface PasteContent',
        });
        expect(editorView.state.doc).toEqualDocument(doc(panel({})(p('abc'))));
      });
      it('the panel should not disappear when pasting multiple paragraphs', () => {
        dispatchPasteEvent(editorView, {
          html: `<meta charset='utf-8'><p data-pm-slice="0 0 []">abc</p><p>123</p>`,
          plain: 'export interface PasteContent',
        });
        expect(editorView.state.doc).toEqualDocument(
          doc(panel({})(p('abc'), p('123'))),
        );
      });
    });
    describe('into an actions list', () => {
      beforeEach(() => {
        const createEditor = createEditorFactory();
        const editor = (doc: DocBuilder) =>
          createEditor({
            doc,
            editorProps: {
              allowPanel: {
                allowCustomPanel: true,
                allowCustomPanelEdit: true,
              },
              allowTasksAndDecisions: true,
            },
          });

        ({ editorView } = editor(
          doc(
            taskList({ localId: 'test-id' })(
              taskItem({ localId: 'test-id' })('{<>}'),
            ),
          ),
        ));
      });

      it('the first action item should not disappear when pasting one paragraph', () => {
        dispatchPasteEvent(editorView, {
          html: `<meta charset='utf-8'><p data-pm-slice="0 0 []">abc</p>`,
          plain: 'export interface PasteContent',
        });
        expect(editorView.state.doc).toEqualDocument(
          doc(
            taskList({ localId: 'test-id' })(
              taskItem({ localId: 'test-id' })('abc'),
            ),
          ),
        );
      });
      it('the first action item should not disappear when pasting multiple paragraphs', () => {
        dispatchPasteEvent(editorView, {
          html: `<meta charset='utf-8'><p data-pm-slice="0 0 []">abc</p><p>123</p>`,
          plain: 'export interface PasteContent',
        });
        expect(editorView.state.doc).toEqualDocument(
          doc(
            taskList({ localId: 'test-id' })(
              taskItem({ localId: 'test-id' })('abc'),
            ),
            p('123'),
          ),
        );
      });
    });
    describe('into a decision list', () => {
      beforeEach(() => {
        const createEditor = createEditorFactory();
        const editor = (doc: DocBuilder) =>
          createEditor({
            doc,
            editorProps: {
              allowPanel: {
                allowCustomPanel: true,
                allowCustomPanelEdit: true,
              },
              allowTasksAndDecisions: true,
            },
          });

        ({ editorView } = editor(
          doc(
            decisionList({ localId: 'test-id' })(
              decisionItem({ localId: 'test-id' })('{<>}'),
            ),
          ),
        ));
      });

      it('the first decision item should not disappear when pasting one paragraph', () => {
        dispatchPasteEvent(editorView, {
          html: `<meta charset='utf-8'><p data-pm-slice="0 0 []">abc</p>`,
          plain: 'export interface PasteContent',
        });
        expect(editorView.state.doc).toEqualDocument(
          doc(
            decisionList({ localId: 'test-id' })(
              decisionItem({ localId: 'test-id' })('abc'),
            ),
          ),
        );
      });
      it('the first decision item should not disappear when pasting multiple paragraphs', () => {
        dispatchPasteEvent(editorView, {
          html: `<meta charset='utf-8'><p data-pm-slice="0 0 []">abc</p><p>123</p>`,
          plain: 'export interface PasteContent',
        });
        expect(editorView.state.doc).toEqualDocument(
          doc(
            decisionList({ localId: 'test-id' })(
              decisionItem({ localId: 'test-id' })('abc'),
            ),
            p('123'),
          ),
        );
      });
    });
  });
});
