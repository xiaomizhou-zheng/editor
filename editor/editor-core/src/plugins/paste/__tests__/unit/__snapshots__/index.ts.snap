// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paste plugin With plainTextPasteLinkification FF disabled cmd+shift+v content with link pastes plain text and doesn't create hyperlink 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "export interface PasteContent https://www.google.com some more content",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF disabled cmd+shift+v content with link preserves current formatting when pasting, doesn't create hyperlink (ie. removes formatting, applies active formatting) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": "export interface PasteContent https://www.google.com some more content",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF disabled cmd+shift+v content without any links pastes plain text (ie. removes formatting from pasted content) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "export interface PasteContent",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF disabled cmd+shift+v content without any links preserves current formatting when pasting (ie. removes formatting, applies active formatting) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": "export interface PasteContent",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF enabled cmd+shift+v content with link pastes plain text but creates hyperlink (ie. removes formatting from pasted content) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "export interface PasteContent ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.google.com",
              },
              "type": "link",
            },
          ],
          "text": "https://www.google.com",
          "type": "text",
        },
        Object {
          "text": " some more content",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF enabled cmd+shift+v content with link preserves current formatting when pasting, creates hyperlink (ie. removes formatting, applies active formatting) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": "export interface PasteContent ",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "attrs": Object {
                "__confluenceMetadata": null,
                "href": "https://www.google.com",
              },
              "type": "link",
            },
            Object {
              "type": "strong",
            },
          ],
          "text": "https://www.google.com",
          "type": "text",
        },
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": " some more content",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF enabled cmd+shift+v content without any links pastes plain text (ie. removes formatting from pasted content) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "export interface PasteContent",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin With plainTextPasteLinkification FF enabled cmd+shift+v content without any links preserves current formatting when pasting (ie. removes formatting, applies active formatting) 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": "export interface PasteContent",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
